tests/test_cases/test_inertial_writes/inertial_writes_tests.py

```diff
@@ -0,0 +1,114 @@
+    else (),
+    skip=SIM_NAME.startswith("riviera") and vhdl and _trust_inertial,
+)
+async def test_writes_on_timer_seen_on_edge(dut):
+    # steady state
+    dut.clk.value = 0
+    await Timer(10, "ns")
+
+    # inertial write on a signal
+    dut.clk.value = 1
+
+    # check we can register an edge trigger on the signal we just changed because it hasn't taken effect yet
+    await with_timeout(RisingEdge(dut.clk), 10, "ns")
+
+
+if _trust_inertial:
+    expect_fail = False
+elif SIM_NAME.startswith(("riviera", "modelsim")) and vhdl:
+    expect_fail = False
+elif SIM_NAME.startswith("verilator"):
+    expect_fail = False
+else:
+    expect_fail = True
+
+
+# Verilator currently only does vpiNoDelay writes, so this works.
+# Riviera and Questa on VHDL designs seem to apply inertial writes in this state immediately,
+# presumably because it's the NBA application region.
+# This test will fail because the ReadWrite write applicator task does inertial writes of it's own.
+@cocotb.test(expect_fail=expect_fail)
+async def test_read_back_in_readwrite(dut):
+    # steady state
+    dut.clk.value = 0
+    await Timer(10, "ns")
+
+    # write in the "normal" phase
+    dut.clk.value = 1
+
+    # assert we can read back what we wrote in the ReadWrite phase
+    await ReadWrite()
+    assert dut.clk.value == 1
+
+
+if not _trust_inertial:
+    expect_fail = False
+elif SIM_NAME.startswith(("icarus", "verilator")):
+    expect_fail = True
+elif SIM_NAME.startswith("modelsim") and verilog:
+    expect_fail = True
+elif SIM_NAME.startswith("xmsim") and vhdl:
+    expect_fail = True
+
+
+# Verilator currently only does vpiNoDelay writes.
+# Icarus, Questa VPI, and Xcelium VHPI inertial writes aren't actually inertial.
+@cocotb.test(expect_fail=expect_fail)
+async def test_writes_dont_update_hdl_this_delta(dut):
+    cocotb.start_soon(Clock(dut.clk, 10, "ns").start())
+
+    # assert steady state
+    dut.stream_in_data.value = 0
+    await RisingEdge(dut.clk)
+    await ReadOnly()
+    assert dut.stream_out_data_registered.value == 0
+
+    # write on the clock edge
+    await RisingEdge(dut.clk)
+    dut.stream_in_data.value = 1
+
+    # ensure that the write data wasn't used on this clock cycle
+    await ReadOnly()
+    assert dut.stream_out_data_registered.value == 0
+
+    # ensure that the write data made it on the result of the next clock cycle
+    await RisingEdge(dut.clk)
+    await ReadOnly()
+    assert dut.stream_out_data_registered.value == 1
+
+
+# Verilator currently only does vpiNoDelay writes.
+# This test will fail because writes scheduled in Timer won't be applied until ReadWrite.
+@cocotb.test(
+    expect_fail=not _trust_inertial
+    or (SIM_NAME.startswith("verilator") and _trust_inertial),
+)
+async def test_write_in_timer_seen_before_readwrite(dut):
+    dut.clk.value = 0
+    await Timer(1, "ns")
+    dut.clk.value = 1
+    t1 = RisingEdge(dut.clk)
+    t2 = ReadWrite()
+    t_res = await First(t1, t2)
+    assert t_res is t1
```

Comment:
I'm becoming suspect of this test. Maybe this is intended? The end of the delta cycle callback should come before the value change on the next delta, right?

Reply:
We use `vhpiCbAfterDelay` callback for `Timer`, so that is executed in step **b.3** of the simulation cycle. In that callback we update the signal value with a deposit.
In step **d**, signals are updated in the order defined in 14.7.4 and deposits are no longer scheduled. This should eventually cause the value change callback to be called before moving to the next step.
In step **h.1**, `vhpiCbRepLastKnownDeltaCycle` callbacks are executed (the `ReadWrite` trigger uses this).

So I think the value change callback should occur before `ReadWrite`.